README.md
# node-gitlog

    Git log parser for Node.JS

[![build status](https://api.travis-ci.org/bluelovers/node-gitlog.svg)](http://travis-ci.org/bluelovers/node-gitlog)
[![dependency status](https://david-dm.org/bluelovers/node-gitlog.svg)](https://david-dm.org/bluelovers/node-gitlog)

## Installation

     npm install gitlog2

### breaking change

#### v5.x

- from `v5.x` change date format, and make `authorDate`/`committerDate` is same format
  (`2019-01-06 04:54:09 +0800`)
- support return `unix timestamp`, so u don't need worry about how to convert date format back to timestamp
- allow `await` when use `callback` mode, but remember u can't change `return value` when use `callback`

## Usage

[API](index.d.ts)

```js
const gitlog = require('gitlog2').gitlog;

const options =
    { repo: __dirname + '/test-repo-folder'
    , number: 20
    , author: 'Dom Harrington'
    , fields:
      [ 'hash'
      , 'abbrevHash'
      , 'subject'
      , 'authorName'
      , 'authorDateRel'
      ]
    , execOptions:
      { maxBuffer: 1000 * 1024
      }
    };

// Asynchronous (with Callback)
gitlog(options, function(error, commits) {
  // Commits is an array of commits in the repo
  console.log(commits)
});

// Synchronous
let commits = gitlog(options);
console.log(commits);
```

## Options

See [git log](http://git-scm.com/docs/git-log)

### repo
The location of the repo, required field.

### number
The number of commits to return, defaults to 10.

### since/after
Show commits more recent than a specific date.

### until/before
Show commits older than a specific date.

### author/committer
Limit the commits output to ones with author/committer header lines that match the specified pattern.

### nameStatus
Below fields was returned from the log:

- files - changed files names (array)
- status - changed files status (array)

This option is enabled by default.

### findCopiesHarder
Much more likely to set status codes to 'C' if files are exact copies of each other.

This option is disabled by default.

### all
Find commits on all branches instead of just on the current one.

This option is disabled by default.

### branch ([revision range](https://git-scm.com/docs/git-log#git-log-ltrevisionrangegt))
Show only commits in the specified branch or revision range.

By default uses the current branch and defaults to `HEAD` (i.e. the whole history leading to the current commit).

### execOptions

Type: `Object`

Specify some options to be passed to the [.exec()](http://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback) method:

- `cwd` String *Current working directory of the child process*
- `env` Object *Environment key-value pairs*
- `setsid` Boolean
- `encoding` String *(Default: 'utf8')*
- `timeout` Number *(Default: 0)*
- `maxBuffer` Number *(Default: 200\*1024)*
- `killSignal` String *(Default: 'SIGTERM')*

### optional fields
An array of fields to return from the log, here are the possible options:

- hash - the long hash of the commit e.g. 7dd0b07625203f69cd55d779d873f1adcffaa84a
- abbrevHash - the abbreviated commit hash e.g. 7dd0b07
- treeHash - the tree hash of the commit
- abbrevTreeHash - the abbreviated commit hash
- parentHashes - the parent hashes
- abbrevParentHashes - the abbreviated parent hashes
- authorName - author name of the commit
- authorEmail - author email of the commit
- authorDate - author date of the commit
- authorDateRel - relative author date of the commit
- committerName - committer name
- committerEmail - committer email
- committerDate - committer date
- committerDateRel - relative committer date
- subject - commit message (first line)
- body - commit body
- rawBody - raw body (subject + body)

Defaults to 'abbrevHash', 'hash', 'subject' and 'authorName'.

## How it works

This module works by executing a child process (using `child_process.exec()`) to the `git` executable, then parsing the stdout into commits. This is done using the `--pretty` command line option which allows you to provide a custom formatter to `git log`. To enable easy parsing the format is delimited by a tab (`\t`) character.

## Example

```javascript
  { hash: '6a7ef5e3b3d9c77743140443c8f9e792b0715721',
    abbrevHash: '6a7ef5e',
    treeHash: 'f1bf51b15b48a00c33727f364afef695029864c0',
    abbrevTreeHash: 'f1bf51b',
    parentHashes: 'cfe06dbdb8d0a193640977e016a04678f8f3b04f',
    abbrevParentHashes: 'cfe06dbdb8d0a193640977e016a04678f8f3b04f',
    authorName: 'Dom Harrington',
    authorEmail: 'dom@harringtonxxxxx',
    authorDate: '2015-04-09 09:39:23 +0100',
    authorDateRel: '6 days ago',
    committerName: 'Dom Harrington',
    committerEmail: 'dom@harringtonxxxxx',
    committerDate: 'Thu Apr 9 09:39:23 2015 +0100',
    committerDateRel: '6 days ago',
    subject: '1.0.0',
    body: '',
    rawBody: '.\n',
    status: [ 'M' ],
    files: [ 'package.json' ],
    fileStatus: [ [ 'M', 'package.json' ] ]}
```


CHANGELOG.md
# Change Log

Please see the individual package changelogs for what's new:

* 🌏 [`@git-lazy/clone`](./packages/@git-lazy/clone/CHANGELOG.md "packages/@git-lazy/clone") *packages/@git-lazy/clone*
* 🌏 [`@git-lazy/clone-subdir`](./packages/@git-lazy/clone-subdir/CHANGELOG.md "packages/@git-lazy/clone-subdir") *packages/@git-lazy/clone-subdir*
* 🌏 [`@git-lazy/const`](./packages/@git-lazy/const/CHANGELOG.md "packages/@git-lazy/const") *packages/@git-lazy/const*
* 🌏 [`@git-lazy/git-dir-path`](./packages/@git-lazy/git-dir-path/CHANGELOG.md "packages/@git-lazy/git-dir-path") *packages/@git-lazy/git-dir-path*
* 🌏 [`@git-lazy/git-env-var`](./packages/@git-lazy/git-env-var/CHANGELOG.md "packages/@git-lazy/git-env-var") *packages/@git-lazy/git-env-var*
* 🌏 [`@git-lazy/git-rebase-todo-parser`](./packages/@git-lazy/git-rebase-todo-parser/CHANGELOG.md "packages/@git-lazy/git-rebase-todo-parser") *packages/@git-lazy/git-rebase-todo-parser*
* 🌏 [`@git-lazy/subtree`](./packages/@git-lazy/subtree/CHANGELOG.md "packages/@git-lazy/subtree") *packages/@git-lazy/subtree*
* 🌏 [`@git-lazy/tag`](./packages/@git-lazy/tag/CHANGELOG.md "packages/@git-lazy/tag") *packages/@git-lazy/tag*
* 🌏 [`@git-lazy/tag-list`](./packages/@git-lazy/tag-list/CHANGELOG.md "packages/@git-lazy/tag-list") *packages/@git-lazy/tag-list*
* 🌏 [`@git-lazy/util-args`](./packages/@git-lazy/util-args/CHANGELOG.md "packages/@git-lazy/util-args") *packages/@git-lazy/util-args*
* 🌏 [`@git-lazy/branch`](./packages/branch/CHANGELOG.md "packages/branch") *packages/branch*
* 🌏 [`@git-lazy/change-root-dir`](./packages/change-root-dir/CHANGELOG.md "packages/change-root-dir") *packages/change-root-dir*
* 🌏 [`@git-lazy/debug`](./packages/debug/CHANGELOG.md "packages/debug") *packages/debug*
* 🌏 [`@git-lazy/diff-staged`](./packages/diff-staged/CHANGELOG.md "packages/diff-staged") *packages/diff-staged*
* 🌏 [`git-decode`](./packages/git-decode/CHANGELOG.md "packages/git-decode") *packages/git-decode*
* 🌏 [`git-diff-from`](./packages/git-diff-from/CHANGELOG.md "packages/git-diff-from") *packages/git-diff-from*
* 🌏 [`git-dummy-commit2`](./packages/git-dummy-commit/CHANGELOG.md "packages/git-dummy-commit") *packages/git-dummy-commit*
* 🌏 [`git-rev-range`](./packages/git-rev-range/CHANGELOG.md "packages/git-rev-range") *packages/git-rev-range*
* 🌏 [`git-root2`](./packages/git-root2/CHANGELOG.md "packages/git-root2") *packages/git-root2*
* 🌏 [`gitlog2`](./packages/gitlog2/CHANGELOG.md "packages/gitlog2") *packages/gitlog2*
* 🌏 [`@git-lazy/info`](./packages/info/CHANGELOG.md "packages/info") *packages/info*
* 🌏 [`@git-lazy/remote`](./packages/remote/CHANGELOG.md "packages/remote") *packages/remote*
* 🌏 [`@git-lazy/root`](./packages/root/CHANGELOG.md "packages/root") *packages/root*
* 🌏 [`@git-lazy/spawn`](./packages/spawn/CHANGELOG.md "packages/spawn") *packages/spawn*
* 🌏 [`@git-lazy/untracked`](./packages/untracked/CHANGELOG.md "packages/untracked") *packages/untracked*
* 🌏 [`@git-lazy/util`](./packages/util/CHANGELOG.md "packages/util") *packages/util*
* 🌏 [`@lazy-spawn/stringify`](./packages/@lazy-spawn/stringify/CHANGELOG.md "packages/@lazy-spawn/stringify") *packages/@lazy-spawn/stringify*
* 🌏 [`@lazy-spawn/strip-ansi`](./packages/@lazy-spawn/strip-ansi/CHANGELOG.md "packages/@lazy-spawn/strip-ansi") *packages/@lazy-spawn/strip-ansi*



lib/type.ts
/**
 * Created by user on 2019/1/6/006.
 */

import { SpawnSyncOptions } from "child_process";

export const defaultFields: IFieldsArray = ['abbrevHash', 'hash', 'subject', 'authorName'];
export const defaultOptions: IOptions = {
	number: 10,
	fields: defaultFields,
	nameStatus: true,
	findCopiesHarder: false,
	all: false,
};

export interface IOptionsGitFlogsExtra
{
	/**
	 * The number of commits to return.
	 */
	number?: number,

	/**
	 * `-m --first-parent` use with firstParent
	 * 讓合併紀錄可以顯示此合併變動的檔案狀態
	 *
	 * This flag makes the merge commits show the full diff like regular commits;
	 * for each merge parent, a separate log entry and diff is generated.
	 * An exception is that only diff against the first parent is shown when --first-parent option is given;
	 * in that case, the output represents the changes the merge brought into the then-current branch.
	 *
	 * @see https://stackoverflow.com/questions/37801342/using-git-log-to-display-files-changed-during-merge
	 */
	displayFilesChangedDuringMerge?: boolean,
}

export interface IOptionsGitFlogs
{
	nameStatus?: boolean,

	/**
	 * Much more likely to set status codes to 'C' if files are exact copies of each other.
	 */
	findCopiesHarder?: boolean,
	/**
	 * Find commits on all branches instead of just on the current one.
	 */
	all?: boolean,

	/**
	 * Print only merge commits. This is exactly the same as --min-parents=2
	 */
	merges?: boolean,
	/**
	 * Do not print commits with more than one parent. This is exactly the same as --max-parents=1
	 */
	noMerges?: boolean,

	/**
	 * 只返回對主分支的提交歷史
	 *
	 * Follow only the first parent commit upon seeing a merge commit. This option can give a better overview when viewing the evolution of a particular topic branch, because merges into a topic branch tend to be only about adjusting to updated upstream from time to time, and this option allows you to ignore the individual commits brought in to your history by such a merge. Cannot be combined with --bisect.
	 */
	firstParent?: boolean,

	/**
	 * Continue listing the history of a file beyond renames
	 * (works only for a single file).
	 */
	follow?: boolean,

	/**
	 * 较为完整的提交历史
	 * Same as the default mode, but does not prune some history.
	 *
	 * @see https://judes.me/tech/2017/02/06/overwrite-code.html
	 */
	fullHistory?: boolean,
	/**
	 * 最为完整的提交历史
	 * 搭配 fullHistory 使用
	 *
	 * All commits in the simplified history are shown.
	 */
	sparse?: boolean,

	/**
	 * Additional option to --full-history to remove some needless merges from the resulting history, as there are no selected commits contributing to this merge.
	 */
	simplifyMerges?: boolean,
}

export interface IOptionsGitWithValue
{
	/**
	 * Skip number commits before starting to show the commit output.
	 */
	skip?: number,

	/**
	 * Show commits more recent than a specific date.
	 */
	since?: string,
	/**
	 * Show commits more recent than a specific date.
	 */
	after?: string,

	/**
	 * Show commits older than a specific date.
	 */
	until?: string,
	/**
	 * Show commits older than a specific date.
	 */
	before?: string,

	/**
	 * Limit the commits output to ones with author/committer header lines that match the specified pattern.
	 */
	committer?: string,
	/**
	 * Limit the commits output to ones with author/committer header lines that match the specified pattern.
	 */
	author?: string,
}

/**
 * @see https://cloud.tencent.com/developer/section/1138655
 */
export type IOptions = IOptionsGitFlogs & IOptionsGitWithValue & IOptionsGitFlogsExtra & {

	/**
	 * An array of fields to return from the log
	 */
	fields?: IFieldsArray,

	/**
	 * The location of the repo, required field.
	 */
	repo?: string,
	cwd?: string,

	nameStatusFiles?: boolean,

	/**
	 * Specify some options to be passed to the .exec() method:
	 */
	execOptions?: SpawnSyncOptions,

	/**
	 * Show only commits in the specified branch or revision range.
	 */
	branch?: string,
	/**
	 * Show only commits that are enough to explain how the files that match the specified paths came to be.
	 */
	file?: string,
	files?: string[],

	returnAllFields?: boolean,

	fnHandleBuffer?(buf: Buffer): string,
};

/**
 * https://ruby-china.org/topics/939
 * https://git-scm.com/docs/pretty-formats
 */
export const fields: {
	[name in keyof IParseCommitCore]: string
} = {
	hash: '%H',
	abbrevHash: '%h',
	treeHash: '%T',
	abbrevTreeHash: '%t',
	parentHashes: '%P',
	abbrevParentHashes: '%P',
	authorName: '%an',
	authorEmail: '%ae',
	authorDate: '%ai',
	authorDateRel: '%ar',
	authorDateUnixTimestamp: '%at',
	committerName: '%cn',
	committerEmail: '%ce',
	committerDate: '%ci',
	committerDateRel: '%cr',
	committerDateUnixTimestamp: '%ct',
	subject: '%s',
	body: '%b',
	rawBody: '%B',
	tags: '%D',

	commitNotes: '%N',

	encoding: '%e',

	refNames: '%D',
};

export interface IParseCommitCore
{
	/**
	 * commit hash
	 */
	hash?: string,
	/**
	 * 缩短的 commit hash
	 */
	abbrevHash?: string,
	treeHash?: string,
	/**
	 * 缩短的 tree hash
	 */
	abbrevTreeHash?: string,
	parentHashes?: string,
	/**
	 * 缩短的 parent hashes
	 */
	abbrevParentHashes?: string,
	authorName?: string,
	authorEmail?: string,
	/**
	 * 日期, ISO 8601 格式
	 * @example 2019-01-06 04:54:09 +0800
	 */
	authorDate?: string,
	/**
	 * 日期, 相对格式(1 day ago)
	 * @example 20 hours ago
	 */
	authorDateRel?: string,
	/**
	 * 日期, UNIX timestamp
	 */
	authorDateUnixTimestamp?: number,
	committerName?: string,
	committerEmail?: string,
	/**
	 * 提交日期, ISO 8601 格式
	 */
	committerDate?: string,
	/**
	 * 提交日期, 相对格式(1 day ago)
	 */
	committerDateRel?: string,
	/**
	 * 提交日期, UNIX timestamp
	 */
	committerDateUnixTimestamp?: number,
	/**
	 * commit 信息标题
	 */
	subject?: string,

	tags?: string[],

	/**
	 * 每個檔案對應的變動狀態
	 * 與 files 內容相對應
	 */
	status?: EnumFileStatus[],
	/**
	 * 每個變動狀態的檔案名稱
	 * 與 status 內容相對應
	 */
	files?: string[],

	/**
	 * commit 信息内容
	 */
	body?: string,
	/**
	 * raw body (subject + body)
	 */
	rawBody?: string,

	commitNotes?: string,

	encoding?: string,

	/**
	 * ref names without the " (", ")" wrapping.
	 * @example HEAD -> master, origin/master
	 */
	refNames?: string,

}

export type IParseCommit = IParseCommitCore & {

	/**
	 * 每個變動狀態的變動狀態與檔案名稱
	 * = status + files
	 */
	fileStatus?: [EnumFileStatus, string][],

	/**
	 * 此 log 出現的原始順序
	 * 作為後期處理時可以額外做判斷的依據
	 */
	_index: number,
}

/**
 * https://git-scm.com/docs/git-status
 */
export enum EnumFileStatus
{
	/**
	 * unmodified
	 */
	UNMODIFIED = ' ',
	/**
	 * modified
	 */
	MODIFIED = 'M',
	/**
	 * added
	 */
	ADDED = 'A',
	/**
	 * deleted
	 */
	DELETED = 'D',
	/**
	 * renamed
	 */
	RENAMED = 'R',
	/**
	 * copied
	 */
	COPIED = 'C',
	/**
	 * updated but unmerged
	 */
	UPDATED_BUT_UNMERGED = 'U',
}

export type IFieldsArray = Array<keyof typeof fields>

/**
 * for moment
 */
export enum EnumGitDateFormat
{
	/**
	 * git style ISO 8601
	 *
	 * @example 2019-01-06 04:54:09 +0800
	 */
	ISO_8601 = 'YYYY-MM-DD HH:mm:ss Z',
	/**
	 * git style RFC 2822
	 */
	RFC_2822 = 'ddd MMM DD HH:mm:ss YYYY ZZ',

	// old version

	AUTHOR_DATE = 'YYYY-MM-DD HH:mm:ss Z',
	COMMITTER_DATE = 'ddd MMM DD HH:mm:ss YYYY ZZ',
}

export type IReturnCommits = IParseCommit[];

export type ICommands = Array<number | string>

export const KEY_ORDER: (keyof IParseCommit)[] = [
	'_index',
	'hash',
	'abbrevHash',
	'treeHash',
	'abbrevTreeHash',
	'parentHashes',
	'abbrevParentHashes',
	'authorName',
	'authorEmail',
	'authorDate',
	'authorDateRel',
	'authorDateUnixTimestamp',
	'committerName',
	'committerEmail',
	'committerDate',
	'committerDateRel',
	'committerDateUnixTimestamp',
	'subject',
	'body',
	'rawBody',
	'tags',
	'refNames',
	'status',
	'files',
	'fileStatus',
];
export const delimiter = "\t";
export const notOptFields = ['status', 'files'];

export const enum EnumPrettyFormatFlags
{
	PRETTY = 'pretty',
	FORMAT = 'format',
}

export const enum EnumPrettyFormatMark
{
	BEGIN = '@begin@',

	DELIMITER = '\t',

	END = '@end@',

	JOIN = '',
}

// @ts-ignore
Object.freeze(exports);
